Cadasta/cadasta-arcgis-admin
# Add Sentry logging for Lambda handler

This PR adds Sentry logging to all Lambda handlers. It adds the `lib/sentry` module, which exports `SentryWrapper`. `SentryWrapper` can be applied to Lambda handlers like:

```javascript
import SentryWrapper from '../lib/sentry';

const lambdaHandler = (event) => {}

export default SentryWrapper.handler(hello);
```

`SentryWrapper` wraps the function that is passed to `handler` in a `try .. catch` statement which catches all unhandled errors in the Lambda handler and logs them with Sentry.

If the `SENTRY_DSN` env variable is not configured (if you run the function locally) the error is thrown instead. 

To receive meaningful stack traces in Sentry, we need to create source maps and upload them as artifacts to Sentry. This is done using the web pack plugin `SentryPlugin`. 

When a build is created with `NODE_ENV=production`, a new release is registered with Sentry and the source maps and minified code are attached as artifacts to that release. The release name is the hash of the most recent commit in the current branch. 

## Additions

- `SentryWrapper` module plus tests. 
- The shell script `api/setup-env`, which reads the required env variables from `.env` (Needs to be provided locally) and sets up the env variables (`RELEASE` and `SENTRY_DSN`) for deployment. 

## Changes

- Add `SentryDsn` and `Release` parameters to the SAM template, so they get injected into the Lambda envs
- Change the `build` command to use `NODE_ENV=development` by default, so we don't register a new release and upload the source maps every time we make a local build. 
- Adds  `devtool: 'source-map'` and `SentryPlugin` to `web pack.config`. They are needed to create source maps and to upload the source maps as artifacts to Sentry whenever we do a production build. 

## How to test

- Create a new build with `NODE_ENV=production` (Follow the README additions to set up your env properly)
- Get a the Sentry DSN for the arcgis_admin project and export it to `SENTRY_DSN` env variable
- Run the local API server
- Go to `localhost:3000/hello`, which throws an error, which should be logged in Sentry after a bit. 

**NOTE:** Please have a thorough look on my use of types etc. and let me know about anything that doesn't comply with best practices.
